src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
  * 02111-1307, USA.
  *)
 
-(** Module [Lwt]: cooperative light-weight threads. *)
 
-(** This module defines {e cooperative light-weight threads} with
-    their primitives. A {e light-weight thread} represent a
-    computation that may be not terminated, for example because it is
-    waiting for some event to happen.
 
-    Lwt threads are cooperative in the sense that switching to another
-    thread is always explicit (with {!wakeup} or {!wakeup_exn}). When a
-    thread is running, it executes as much as possible, and then
-    returns (a value or an error) or sleeps.
+(** Asynchronous programming with promises.
 
-    Note that inside a Lwt thread, exceptions must be raised with
-    {!fail} instead of [raise]. Also the [try ... with ...]
-    construction will not catch Lwt errors. You must use {!catch}
-    instead. You can also use {!wrap} for functions that may raise
-    normal exception.
+    {b Promises} are placeholders for values that take a long time to compute.
```

Comment:
not all promised values take long time to compute. So I would suggest the following wording: "Promises are placeholders for values that are determined asynchronously". Also, your definition here doesn't take into account that a promise might be rejected. 

See also ours definitions of [futures][1], that is another name for what you're calling promises :)

[1]: https://github.com/BinaryAnalysisPlatform/bap/blob/master/lib/bap_future/bap_future.mli#L23

Reply:
What happened here is that I gave up hope trying to find a formulation that is both accessible and precise for the first sentence. Of course, what is written there is not fully right. But:

- The word "asynchronous" seems like it will require further definition, and I want a sentence comprehensible immediately here.
- "Promises are placeholders for values that are determined asynchronously" is also imprecise, in a slightly different way, because promises can be created already-determined, for example.

So far, given the set of not-really-correct choices (including some other ones I came up with), I'm trying to choose the incorrect one that gets the most benefit out of being incorrect, by giving as immediate an intuition as possible :) After all, promises are *mostly* to be used as placeholders for values that *might* take some large amount of time to compute :)

As for rejection, I decided to simply leave it out, as a complication to be discussed later. It doesn't seem that different from saying that a value of type `int -> string` in OCaml is a function that evaluates to `string`s, and forgetting to mention that it might raise an exception (or not terminate).

Reply:
Also, this reference is suffering the effects of having the tutorial inlined. Once that's factored out, I think it might be more realistic to use your suggestion, be precise about rejection in the first paragraph immediately, etc., and refer readers to the tutorial for a more progressive introduction. However, since I don't have the separate tutorial now to see how that will look, I prefer to introduce things progressively in `lwt.mli` for the time being.